signature/jws/plugin.go

```diff
@@ -0,0 +1,195 @@
+	}
+	return resp, nil
+}
+
+func (s *PluginSigner) generateSignature(ctx context.Context, desc signature.Descriptor, opts notation.SignOptions) ([]byte, error) {
+	config := s.mergeConfig(opts.PluginConfig)
+	// Get key info.
+	key, err := s.describeKey(ctx, config)
+	if err != nil {
+		return nil, err
+	}
+
+	// Check keyID is honored.
+	if s.KeyID != key.KeyID {
+		return nil, fmt.Errorf("keyID mismatch")
+	}
+
+	// Get algorithm associated to key.
+	alg := key.KeySpec.SignatureAlgorithm()
+	if alg == "" {
+		return nil, fmt.Errorf("keySpec %q not supported: ", key.KeySpec)
+	}
+
+	// Generate payload to be signed.
+	payload := packPayload(desc, opts)
+	if err := payload.Valid(); err != nil {
+		return nil, err
+	}
+
+	// Generate signing string.
+	token := jwtToken(alg.JWS(), payload)
+	signing, err := token.SigningString()
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal signing payload: %v", err)
+	}
+
+	// Execute plugin sign command.
+	req := &plugin.GenerateSignatureRequest{
+		ContractVersion: plugin.ContractVersion,
+		KeyID:           s.KeyID,
+		KeySpec:         key.KeySpec,
+		Hash:            alg.Hash(),
+		Payload:         []byte(signing),
+		PluginConfig:    config,
+	}
+	out, err := s.Runner.Run(ctx, plugin.CommandGenerateSignature, req)
+	if err != nil {
+		return nil, fmt.Errorf("generate-signature command failed: %w", err)
+	}
+	resp, ok := out.(*plugin.GenerateSignatureResponse)
+	if !ok {
+		return nil, fmt.Errorf("plugin runner returned incorrect generate-signature response type '%T'", out)
+	}
+
+	// Check keyID is honored.
+	if s.KeyID != resp.KeyID {
+		return nil, fmt.Errorf("keyID mismatch")
+	}
+
+	// Check algorithm is supported.
+	jwsAlg := resp.SigningAlgorithm.JWS()
+	if jwsAlg == "" {
+		return nil, fmt.Errorf("signing algorithm %q not supported", resp.SigningAlgorithm)
+	}
+
+	// Check certificate chain is not empty.
+	if len(resp.CertificateChain) == 0 {
+		return nil, errors.New("empty certificate chain")
+	}
+
+	certs, err := parseCertChain(resp.CertificateChain)
+	if err != nil {
+		return nil, err
+	}
+
+	// Verify the hash of the request payload against the response signature
+	// using the public key of the signing certificate.
+	// At this point, resp.Signature is not base64-encoded,
+	// but verifyJWT expects a base64URL encoded string.
+	signed64Url := base64.RawURLEncoding.EncodeToString(resp.Signature)
+	err = verifyJWT(jwsAlg, signing, signed64Url, certs[0])
+	if err != nil {
+		return nil, fmt.Errorf("verification error: %v", err)
+	}
+
+	// Check the the certificate chain conforms to the spec.
+	if err := verifyCertExtKeyUsage(certs[0], x509.ExtKeyUsageCodeSigning); err != nil {
+		return nil, fmt.Errorf("signing certificate does not meet the minimum requirements: %w", err)
+	}
+
+	// Assemble the JWS signature envelope.
+	return jwtEnvelope(ctx, opts, signing+"."+signed64Url, resp.CertificateChain)
```

Comment:
Method name should be `jwsEnvelope` , `converttoJws` or similar.

Reply:
We should including a timestamp signature similar to [jws signer](https://github.com/notaryproject/notation-go/blob/main/signature/jws/signer.go#L126-L132).

Reply:
`jwtEnvelope` already includes the timestamp signature if the appropriate signing option is set. Is there anything missing?

Reply:
You are right, the renamed `jwsEnvelope()` will timestamp both in the plugin and non plugin flow.